lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp

```diff
@@ -2523,59 +2549,107 @@ bool EliminateSliceConcat::run(Function *F, const CompilationContext &cctx) {
     // 2) merging Slices from different sources
     // e.g. Slice(A1)-Slice(B1)-Slice(A2)-Slice(B2), A1 and A2, B1 and B2 are
     // consecutive respectively
-    std::vector<std::vector<SliceNode *>> consecutiveSlices;
+    //
+    // Store consecutive slices along *any* dimension. If the consecutive
+    // slices' dimension is the same as the concat, the concat can be removed.
+    // If the slices' dimension is different from the concat, the nodes
+    // can be replaced with a single slice+reshape.
+    std::vector<std::pair<unsigned_t /* dimension of slicing */,
+                          std::vector<SliceNode *>>> consecutiveSlices;
     std::vector<SliceNode *> currConsecutiveSlices;
     SliceNode *lastSN = nullptr;
+    unsigned_t lastDim = -1;
     for (auto &concatInput : CN->getInputs()) {
       auto *SN = dyn_cast<SliceNode>(concatInput.getNode());
       // slices with multiple users will not be considered
       if (!SN || SN->getResult().getNumUsers() > 1) {
         if (currConsecutiveSlices.size()) {
-          consecutiveSlices.emplace_back(currConsecutiveSlices);
+          consecutiveSlices.emplace_back(lastDim, currConsecutiveSlices);
           currConsecutiveSlices.clear();
         }
+        lastDim = -1;
         lastSN = nullptr;
         continue;
       }
       // slices with different sources will not be considered
+      unsigned_t dim = -1;
       if (lastSN && (lastSN->getInput() != SN->getInput() ||
-                     !areSlicesConsecutive(lastSN, SN, CN->getDim()))) {
-        consecutiveSlices.emplace_back(currConsecutiveSlices);
+                     !findConsecutiveSliceDim(lastSN, SN, &dim))) {
+        consecutiveSlices.emplace_back(lastDim, currConsecutiveSlices);
         currConsecutiveSlices.clear();
       }
+      lastDim = dim;
       lastSN = SN;
       currConsecutiveSlices.emplace_back(SN);
     }
     if (currConsecutiveSlices.size()) {
-      consecutiveSlices.emplace_back(currConsecutiveSlices);
+      consecutiveSlices.emplace_back(lastDim, currConsecutiveSlices);
     }
 
     // Mapping from old Slices to new Slices where the range of each old Slice
     // is a subset of the corresponding new Slice
     std::unordered_map<SliceNode *, SliceNode *> oldToNewSlices;
-    for (const auto &slices : consecutiveSlices) {
+
+    // Mapping from new fused slices to reshape nodes that maybe be needed in
+    // case the fused slices need to be reshaped before concat. This happens
+    // when consecutive slices dimension differs from concat dimension by 1.
+    std::unordered_map<SliceNode *, ReshapeNode *> newSlicesToReshape;
```

Comment:
Just thought of an alternate approach to create a single mapping as such:
```c++
std::unordered_map<SliceNode *, NodeValue *> oldToNewSlicesOrReshape;
```
and similarly, change the lookup logic. This would have a bit more casting but will use less memory (single hashmap vs. two). 

If warranted, I'll upload the new patch to have that change.

Reply:
Done. Much cleaner code. It uses `std::unordered_map<SliceNode *, Node *>`.